doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,433 @@
+|  // remaining API is unchanged                    |
++---------------------------------------------------+
+                          ^
+                         /|\
+                          |
+       +------------------+--------------------------+
+       |template<T, Capacity>                        |
+       |vector : public vector<Type>                 |
+       |                                             |
+       |  (#1)                                       |
+       |  - vector(const vector<T> &)                |
+       |  - vector(vector<T> &&)                     |
+       |  - vector& operator=(const vector<T> &)     |
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based version of cxx containers will not have constructors similar
+to the STL. The only provided constructor has one argument which specifies the
+type of allocator one would like to use for that container. If no argument is
+provided the heap allocator will be used by default.
+
+The constructors and operations defined in (#1) are added to support implicit
+conversion from the generic `cxx::vector` which is required to the stack version
+of a vector. It makes the following operations possible
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int> v1;
+vector<int, 20> v2;
+
+// this works thanks to implicit conversions from vector<int>
+f(v1);
+v2 = v1;
+```
+
+Since every stack based vector is a child of the more generic vector we can with
+the help of (#1) also assign and use stack based vectors of different sizes
+to each other or use them in functions. This allows us to write code like this:
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int, 10> v1;
+vector<int, 20> v2;
+
+// this works since vector<int, 10> is also a vector<int> and can be converted
+v2 = v1;
+v1 = v2; // this can fail when v1 capacity is smaller then v2.size, previously 
+         // it was only possible when v1.capacity < v2.capacity
+f(v1);
+f(v2);
+```
+
+The only restriction are functions which require a stack based vector of a 
+specific size as non const reference. This is usally the case when some function
+would like to set the contents of a vector.
+```cpp
+void setVectorContents(vector<int, 20> &a) {}
+
+vector<int, 10> v1;
+vector<int> v2;
+
+// both calls will result in a compile time warning
+setVectorContents(v1);
+setVectorContents(v2);
+```
+
+### Usage
+
+```cpp
+cxx::vector<int, 20> a;
+a.emplace_back(123); // can be used without a preceding reserve call
+
+cxx::vector<int> c(heapAllocator);
+c.reserve(42); // reserve required
```

Comment:
what should happen if no reserve was called?

Reply:
In the case of the static vector it is not required and nothing will happen. For all other vectors `cxx::Expects` is called with the expectation that reserve was called before - if not `cxx::Expects` will call `std::terminate()` but this can be configured to something else later when we have a more fine grained error handling concept.

Reply:
Can reserve fail?

Reply:
The allocator returns an error via expected which is forwarded to the user when reserve fails.